File: src/specter/sim/paths.py
"""Wave 3 — Path / waypoint geometry for Free Play.

Mirrors `ui/packages/sim-core/src/paths.ts`. Math constants must stay
byte-equivalent across the two languages so the Composer's "default path
for preset" produces identical trajectories in TS-driven UI runs and
Python eval scenarios.

A `Path` is an ordered tuple of `(x, y)` waypoints plus a `closed` flag.
`bounce` is only meaningful for open paths (the follower reverses direction
at endpoints).

`path_follower(path, *, speed, arrival_threshold_m=0.2)` returns a
callable `planner(agents, t, dt)` that yields `{agent_id: (vx, vy)}`
velocity commands.  Per-agent waypoint-index state lives in the closure.
"""

from __future__ import annotations

import math
from collections.abc import Callable, Iterable, Sequence
from dataclasses import dataclass
from typing import Literal, Protocol

Point2 = tuple[float, float]
PathPreset = Literal["LOOP", "LINEAR", "FIGURE-8"]

DEFAULT_ARRIVAL_M = 0.2
FIGURE8_DEFAULT_SAMPLES = 32
LOOP_INSET_FRACTION = 0.2


@dataclass(frozen=True)
class Path:
    waypoints: tuple[Point2, ...]
    closed: bool
    bounce: bool = False


@dataclass(frozen=True)
class Bounds:
    min_x: float
    min_y: float
    max_x: float
    max_y: float


class _AgentLike(Protocol):
    id: str
    x: float
    y: float


def loop_waypoints(points: Iterable[Point2]) -> Path:
    """Closed loop through the given waypoints."""
    return Path(waypoints=tuple((float(p[0]), float(p[1])) for p in points), closed=True)


def linear_path(start: Point2, end: Point2, *, bounce: bool = False) -> Path:
    """Open path from `start` to `end`. With `bounce=True`, the follower
    reverses direction at either endpoint."""
    return Path(
        waypoints=((float(start[0]), float(start[1])), (float(end[0]), float(end[1]))),
        closed=False,
        bounce=bounce,
    )


def figure8(*, center: Point2, a: float, b: float, samples: int = FIGURE8_DEFAULT_SAMPLES) -> Path:
    """Lissajous figure-8: x = cx + a·sin(t), y = cy + b·sin(2t),
    t evenly spaced over [0, 2π) at `samples` points. Closed."""
    cx, cy = center
    pts: list[Point2] = []
    for i in range(samples):
        t = 2.0 * math.pi * i / samples
        pts.append((cx + a * math.sin(t), cy + b * math.sin(2.0 * t)))
    return Path(waypoints=tuple(pts), closed=True)


def freehand(points: Iterable[Point2]) -> Path:
    """Open polyline.  The follower halts at the last waypoint."""
    return Path(waypoints=tuple((float(p[0]), float(p[1])) for p in points), closed=False)


def default_path_for(preset: PathPreset, bounds: Bounds) -> Path:
    """Built-in preset path geometry derived from world bounds.

    LOOP → inset regular pentagon (5 vertices, 20% inset).
    LINEAR → bounds.min → bounds.max diagonal.
    FIGURE-8 → centered Lissajous (a=width/3, b=height/3, 32 samples).
    FREEHAND raises — caller must supply explicit waypoints.
    """
    width = bounds.max_x - bounds.min_x
    height = bounds.max_y - bounds.min_y
    cx = (bounds.min_x + bounds.max_x) / 2.0
    cy = (bounds.min_y + bounds.max_y) / 2.0
    if preset == "LOOP":
        rx = (width / 2.0) * (1.0 - LOOP_INSET_FRACTION)
        ry = (height / 2.0) * (1.0 - LOOP_INSET_FRACTION)
        pts: list[Point2] = []
        for i in range(5):
            theta = -math.pi / 2.0 + 2.0 * math.pi * i / 5.0
            pts.append((cx + rx * math.cos(theta), cy + ry * math.sin(theta)))
        return Path(waypoints=tuple(pts), closed=True)
    if preset == "LINEAR":
        return linear_path((bounds.min_x, bounds.min_y), (bounds.max_x, bounds.max_y))
    if preset == "FIGURE-8":
        return figure8(
            center=(cx, cy),
            a=width / 3.0,
            b=height / 3.0,
            samples=FIGURE8_DEFAULT_SAMPLES,
        )
    raise ValueError(
        f'default_path_for: preset "{preset}" has no default geometry — supply explicit waypoints'
    )


Planner = Callable[[Sequence[_AgentLike], float, float], dict[str, Point2]]


def path_follower(path: Path, *, speed: float, arrival_threshold_m: float = DEFAULT_ARRIVAL_M) -> Planner:
    """Per-agent path-follower planner.

    Returns a callable `(agents, t, dt) -> {id: (vx, vy)}` that commands
    velocity toward each agent's current target waypoint, advancing when
    within `arrival_threshold_m`.  Closed paths wrap to W0; open + bounce
    reverses direction at endpoints; open without bounce halts at the
    last waypoint.
    """
    state: dict[str, list[int]] = {}  # id → [idx, direction]
    waypoints = path.waypoints
    last = len(waypoints) - 1

    def _advance(s: list[int]) -> None:
        if path.closed:
            s[0] = (s[0] + 1) % len(waypoints)
            return
        if path.bounce:
            nxt = s[0] + s[1]
            if nxt < 0 or nxt > last:
                s[1] = -s[1]
                s[0] = s[0] + s[1]
            else:
                s[0] = nxt
            return
        if s[0] < last:
            s[0] += 1

    def planner(agents: Sequence[_AgentLike], _t: float, _dt: float) -> dict[str, Point2]:
        cmds: dict[str, Point2] = {}
        if not waypoints:
            return cmds
        for a in agents:
            s = state.setdefault(a.id, [0, 1])
            target = waypoints[s[0]]
            dx = target[0] - a.x
            dy = target[1] - a.y
            dist = math.hypot(dx, dy)
            if dist <= arrival_threshold_m:
                halted = (not path.closed) and (not path.bounce) and s[0] == last
                _advance(s)
                if halted:
                    cmds[a.id] = (0.0, 0.0)
                    continue
                nxt = waypoints[s[0]]
                ndx = nxt[0] - a.x
                ndy = nxt[1] - a.y
                ndist = math.hypot(ndx, ndy)
                if ndist == 0.0:
                    cmds[a.id] = (0.0, 0.0)
                else:
                    cmds[a.id] = ((ndx / ndist) * speed, (ndy / ndist) * speed)
                continue
            cmds[a.id] = ((dx / dist) * speed, (dy / dist) * speed)
        return cmds

    return planner

File: ui/packages/sim-core/src/paths.ts
/**
 * Wave 3 — Path / waypoint geometry for Free Play.
 *
 * Paths sit *under* flight patterns: a `Path` is an ordered sequence of
 * waypoints with a `closed` flag (wrap to W0 at the last waypoint) and an
 * optional `bounce` flag (reverse direction at endpoints, only meaningful on
 * open paths). `pathFollower(path, opts)` returns a sim-core `Planner` that
 * commands per-agent velocity toward the current waypoint and advances when
 * within `arrivalThresholdM`.
 *
 * Math constants here must stay byte-equivalent to `src/specter/sim/paths.py`
 * — the Wave 5 Composer reads `defaultPathFor(preset, bounds)` from both
 * languages and the lessons need identical trajectories.
 */
import type { Agent } from "./agent";
import type { Planner } from "./scenario";
import type { VelocityCommand } from "./swarm";

export type Point2 = readonly [number, number];

export type Path = {
  waypoints: ReadonlyArray<Point2>;
  closed: boolean;
  bounce?: boolean;
};

export type PathPreset = "LOOP" | "LINEAR" | "FIGURE-8";

export type Bounds = {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
};

const DEFAULT_ARRIVAL_M = 0.2;
const FIGURE8_DEFAULT_SAMPLES = 32;
const LOOP_INSET_FRACTION = 0.2;

/** Closed loop through the given waypoints. */
export function loopWaypoints(points: ReadonlyArray<Point2>): Path {
  return { waypoints: points.map((p) => [p[0], p[1]] as const), closed: true };
}

/** Open path from `start` to `end`. With `bounce: true`, the follower reverses
 * direction when it reaches either endpoint. */
export function linearPath(
  start: Point2,
  end: Point2,
  opts?: { bounce?: boolean },
): Path {
  return {
    waypoints: [
      [start[0], start[1]],
      [end[0], end[1]],
    ],
    closed: false,
    bounce: opts?.bounce ?? false,
  };
}

/** Lissajous figure-8: `x = cx + a·sin(t)`, `y = cy + b·sin(2t)`, t evenly
 * spaced over [0, 2π) at `samples` points. Closed. */
export function figure8(spec: {
  center: Point2;
  a: number;
  b: number;
  samples?: number;
}): Path {
  const samples = spec.samples ?? FIGURE8_DEFAULT_SAMPLES;
  const [cx, cy] = spec.center;
  const waypoints: Point2[] = [];
  for (let i = 0; i < samples; i++) {
    const t = (2 * Math.PI * i) / samples;
    waypoints.push([cx + spec.a * Math.sin(t), cy + spec.b * Math.sin(2 * t)]);
  }
  return { waypoints, closed: true };
}

/** Open polyline. The follower halts (zero velocity) at the last waypoint. */
export function freehand(points: ReadonlyArray<Point2>): Path {
  return { waypoints: points.map((p) => [p[0], p[1]] as const), closed: false };
}

/** Built-in preset path geometry derived from world bounds.
 *  - LOOP: inset regular pentagon (5 vertices, 20% inset)
 *  - LINEAR: bounds.min → bounds.max diagonal
 *  - FIGURE-8: centered Lissajous, a = width/3, b = height/3, 32 samples
 * FREEHAND throws — caller must supply explicit waypoints. */
export function defaultPathFor(preset: PathPreset, bounds: Bounds): Path {
  const width = bounds.maxX - bounds.minX;
  const height = bounds.maxY - bounds.minY;
  const cx = (bounds.minX + bounds.maxX) / 2;
  const cy = (bounds.minY + bounds.maxY) / 2;
  if (preset === "LOOP") {
    const rx = (width / 2) * (1 - LOOP_INSET_FRACTION);
    const ry = (height / 2) * (1 - LOOP_INSET_FRACTION);
    const waypoints: Point2[] = [];
    for (let i = 0; i < 5; i++) {
      // Pentagon vertex i: start at top (theta = -π/2) and go CCW.
      const theta = -Math.PI / 2 + (2 * Math.PI * i) / 5;
      waypoints.push([cx + rx * Math.cos(theta), cy + ry * Math.sin(theta)]);
    }
    return { waypoints, closed: true };
  }
  if (preset === "LINEAR") {
    return linearPath([bounds.minX, bounds.minY], [bounds.maxX, bounds.maxY]);
  }
  if (preset === "FIGURE-8") {
    return figure8({
      center: [cx, cy],
      a: width / 3,
      b: height / 3,
      samples: FIGURE8_DEFAULT_SAMPLES,
    });
  }
  throw new Error(`defaultPathFor: preset "${preset}" has no default geometry — supply explicit waypoints`);
}

type FollowerState = { idx: number; direction: 1 | -1 };

/** Per-agent path-follower Planner.
 *  - Closed paths wrap idx → 0 after the last waypoint.
 *  - Open + bounce reverses direction at endpoints.
 *  - Open without bounce commands zero velocity once the last waypoint is
 *    reached (so the agent stops cleanly instead of overshooting). */
export function pathFollower(
  path: Path,
  opts: { speed: number; arrivalThresholdM?: number },
): Planner {
  const speed = opts.speed;
  const arrival = opts.arrivalThresholdM ?? DEFAULT_ARRIVAL_M;
  const state = new Map<string, FollowerState>();
  const waypoints = path.waypoints;
  const last = waypoints.length - 1;

  const advance = (s: FollowerState): void => {
    if (path.closed) {
      s.idx = (s.idx + 1) % waypoints.length;
      return;
    }
    if (path.bounce) {
      const next = s.idx + s.direction;
      if (next < 0 || next > last) {
        s.direction = (s.direction === 1 ? -1 : 1) as 1 | -1;
        s.idx = s.idx + s.direction;
      } else {
        s.idx = next;
      }
      return;
    }
    if (s.idx < last) s.idx += 1;
  };

  return (agents: ReadonlyArray<Agent>): Record<string, VelocityCommand> => {
    const cmds: Record<string, VelocityCommand> = {};
    if (waypoints.length === 0) return cmds;
    for (const a of agents) {
      if (a.phantom) continue;
      let s = state.get(a.id);
      if (!s) {
        s = { idx: 0, direction: 1 };
        state.set(a.id, s);
      }
      const target = waypoints[s.idx]!;
      const dx = target[0] - a.x;
      const dy = target[1] - a.y;
      const dist = Math.hypot(dx, dy);
      if (dist <= arrival) {
        const halted = !path.closed && !path.bounce && s.idx === last;
        advance(s);
        if (halted) {
          cmds[a.id] = { vx: 0, vy: 0 };
          continue;
        }
        const next = waypoints[s.idx]!;
        const ndx = next[0] - a.x;
        const ndy = next[1] - a.y;
        const ndist = Math.hypot(ndx, ndy);
        if (ndist === 0) {
          cmds[a.id] = { vx: 0, vy: 0 };
        } else {
          cmds[a.id] = { vx: (ndx / ndist) * speed, vy: (ndy / ndist) * speed };
        }
        continue;
      }
      cmds[a.id] = { vx: (dx / dist) * speed, vy: (dy / dist) * speed };
    }
    return cmds;
  };
}
